### PR TITLE
[messages] New framework in iOS 10.0 beta1

### PR DESCRIPTION
**Pull request by VincentDondain**
_Note: I could not build because of https://bugzilla.xamarin.com/show_bug.cgi?id=41786. This PR is for review only._

### CODE REVIEW

**Comment by xamarin-release-manager** [comment]
Build failure

File: src/messages.cs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+		MSSession Session { get; }
+
+		[Export ("senderParticipantIdentifier")]
+		NSUuid SenderParticipantIdentifier { get; }
+
+		[NullAllowed, Export ("layout", ArgumentSemantic.Copy)]
+		MSMessageLayout Layout { get; set; }
+
+		[NullAllowed, Export ("URL", ArgumentSemantic.Copy)]
+		NSUrl URL { get; set; }
```
Comment (dalexsoto) [inline src/messages.cs]: Change to `Url` (UpperCamelCase)

File: src/messages.cs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+		string TrailingCaption { get; set; }
+
+		[NullAllowed, Export ("trailingSubcaption")]
+		string TrailingSubcaption { get; set; }
+
+		[NullAllowed, Export ("image", ArgumentSemantic.Strong)]
+		UIImage Image { get; set; }
+
+		[NullAllowed, Export ("mediaFileURL", ArgumentSemantic.Copy)]
+		NSUrl MediaFileURL { get; set; }
```
Comment (dalexsoto) [inline src/messages.cs]: same `MediaFileUrl` (UpperCamelCase)

File: src/messages.cs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+	}
+
+	[Introduced (PlatformName.iOS, 10, 0)]
+	[BaseType (typeof(NSObject))]
+	[DisableDefaultCtor]
+	interface MSSticker
+	{
+		[Export ("initWithContentsOfFileURL:localizedDescription:error:")]
+		[DesignatedInitializer]
+		IntPtr Constructor (NSUrl fileURL, string localizedDescription, [NullAllowed] out NSError error);
```
Comment (dalexsoto) [inline src/messages.cs]: same, `fileUrl`

File: src/messages.cs [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+
+		[Export ("imageFileURL", ArgumentSemantic.Strong)]
+		NSUrl ImageFileURL { get; }
```
Comment (dalexsoto) [inline src/messages.cs]: same `ImageFileUrl`

File: src/messages.cs [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+		MSStickerSize StickerSize { get; }
+
+		[NullAllowed, Export ("dataSource", ArgumentSemantic.Weak)]
+		IMSStickerBrowserViewDataSource DataSource { get; set; }
+
+		[Export ("reloadData")]
+		void ReloadData ();
+	}
+
+	[BaseType (typeof(UIViewController))]
```
Comment (dalexsoto) [inline src/messages.cs]: missing `[Introduced (PlatformName.iOS, 10, 0)]`??

**Comment by spouliot** [comment]
build

File: src/messages.cs [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+	interface MSMessagesAppViewController
+	{
+		[NullAllowed, Export ("activeConversation", ArgumentSemantic.Strong)]
+		MSConversation ActiveConversation { get; }
+
+		[Export ("presentationStyle", ArgumentSemantic.Assign)]
+		MSMessagesAppPresentationStyle PresentationStyle { get; }
+
+		[Export ("requestPresentationStyle:")]
+		void RequestPresentationStyle (MSMessagesAppPresentationStyle presentationStyle);
```
Comment (spouliot) [inline src/messages.cs]: Maybe `Request` ? The parameter (enum) type will already make it clear it's a presentation style when reading the code (so we can drop the duplication without losing context)

File: src/messages.cs [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+		void InsertMessage (MSMessage message, [NullAllowed] string changeDescription, [NullAllowed] Action<NSError> completionHandler);
+
+		[Export ("insertSticker:completionHandler:")]
+		void InsertSticker (MSSticker sticker, [NullAllowed] Action<NSError> completionHandler);
+
+		[Export ("insertText:completionHandler:")]
+		void InsertText (string text, [NullAllowed] Action<NSError> completionHandler);
+
+		[Export ("insertAttachment:withAlternateFilename:completionHandler:")]
+		void InsertAttachment (NSUrl URL, [NullAllowed] string filename, [NullAllowed] Action<NSError> completionHandler);
```
Comment (spouliot) [inline src/messages.cs]: parameter name should be `url`.

File: src/messages.cs [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+		NSUuid LocalParticipantIdentifier { get; }
+
+		[Export ("remoteParticipantIdentifiers")]
+		NSUuid[] RemoteParticipantIdentifiers { get; }
+
+		[NullAllowed, Export ("selectedMessage")]
+		MSMessage SelectedMessage { get; }
+
+		[Export ("insertMessage:localizedChangeDescription:completionHandler:")]
+		void InsertMessage (MSMessage message, [NullAllowed] string changeDescription, [NullAllowed] Action<NSError> completionHandler);
```
Comment (spouliot) [inline src/messages.cs]: The completion handler (here and the next API) makes it looks like it's missing an `[Async]` ?

**Comment by xamarin-release-manager** [comment]
Build failure

File: src/messages.cs [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+	interface IMSStickerBrowserViewDataSource { }
+
+	[Introduced (PlatformName.iOS, 10, 0)]
+	[Protocol, Model]
+	[BaseType (typeof(NSObject))]
+	interface MSStickerBrowserViewDataSource
+	{
+		[Abstract]
+		[Export ("numberOfStickersInStickerBrowserView:")]
+		nint NumberOfStickers (MSStickerBrowserView stickerBrowserView);
```
Comment (spouliot) [inline src/messages.cs]: Add a `Get` prefix (verb for method)

File: src/messages.cs [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+
+		[Abstract]
+		[Export ("stickerBrowserView:stickerAtIndex:")]
+		MSSticker StickerBrowserView (MSStickerBrowserView stickerBrowserView, nint index);
```
Comment (spouliot) [inline src/messages.cs]: same (Get)

**Comment by xamarin-release-manager** [comment]
Build failure

**Comment by xamarin-release-manager** [comment]
Build success

**Comment by VincentDondain** [comment]
Introspection tests passed just fine.

**Comment by xamarin-release-manager** [comment]
Build success

**Comment by VincentDondain** [comment]
Don't link tests pass.

**Comment by spouliot** [comment]
build

it's the introspection (split form dontlink) tests that are the most important and they crashed in the previous build

**Comment by xamarin-release-manager** [comment]
Build success

**Comment by spouliot** [comment]
@VincentDondain introspection-iso reports several failures, please only merge once they are fixed.

**Comment by xamarin-release-manager** [comment]
Build success

File: src/messages.cs [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+		Large
+	}
+
+	[iOS (10,0)]
+	[BaseType (typeof(UIViewController))]
+	interface MSMessagesAppViewController
+	{
+		// inlined ctor
+		[Export ("initWithNibName:bundle:")]
+		[PostGet ("NibBundle")]
```
Comment (spouliot) [inline src/messages.cs]: That `[PostGet]` is not needed anymore since we have deprecated/removed non-NRC builds.

**Comment by xamarin-release-manager** [comment]
Build success

**Comment by xamarin-release-manager** [comment]
Build success

**Pull request merged**